src/algorithms/all_paths.c

```diff
@@ -86,16 +88,24 @@ Path *AllPathsCtx_NextPath(AllPathsCtx *ctx) {
 			if(depth < ctx->maxLen && !frontierAlreadyOnPath) {
 				// Get frontier neighbors.
 				for(int i = 0; i < ctx->relationCount; i++) {
-					Graph_GetNodeEdges(ctx->g, &frontierNode, ctx->dir, ctx->relationIDs[i],
-									   &ctx->neighbors);
+					Graph_GetNodeEdges(ctx->g, &frontierNode, ctx->dir, ctx->relationIDs[i], &ctx->neighbors);
 				}
 
 				// Add unvisited neighbors to next level.
 				uint32_t neighborsCount = array_len(ctx->neighbors);
 				for(uint32_t i = 0; i < neighborsCount; i++) {
-					// TODO might need to check direction!
+					// Don't follow the frontier edge again.
+					if(frontierId == ENTITY_GET_ID(ctx->neighbors + i)) continue;
 					Node neighbor;
-					if(ctx->dir == GRAPH_EDGE_DIR_OUTGOING) {
+					bool neighbor_is_dest = false;
+					if(ctx->dir == GRAPH_EDGE_DIR_BOTH) {
+						/* If we're traversing bidirectionally, the frontier node will always be the
+						 * source or destination of each edge. We are only interested in the other node. */
+						EntityID destId = Edge_GetDestNodeID(ctx->neighbors + i);
+						neighbor_is_dest = ENTITY_GET_ID(&frontierNode) != destId;
```

Comment:
What if `frontierNode` is connected to itself?
`CREATE (a), (a)-[:R]>(a)`
I believe variable length allows closing a cycle.

One possibility when dealing with bi-directional edge is to call `Graph_GetNodeEdges` twice, once for each direction, that way we'll always know at which end of the edge `frontierNode` is at.